src/asyncresults.jl

```diff
@@ -0,0 +1,256 @@
+
+    "Whether or not the query should be cancelled, if running"
+    should_cancel::Bool
+
+    "Task which errors or returns a LibPQ.jl Result which is created once available"
+    result_task::Task
+
+    function AsyncResult(jl_conn::Connection, result_kwargs::Ref)
+        async_result = new(jl_conn, result_kwargs, false)
+        jl_conn.async_result = async_result
+        return async_result
+    end
+end
+
+function AsyncResult(jl_conn::Connection; kwargs...)
+    return AsyncResult(jl_conn, Ref(kwargs))
+end
+
+function Base.show(io::IO, async_result::AsyncResult)
+    status = if isready(async_result)
+        if iserror(async_result)
+            "errored"
+        else
+            "finished"
+        end
+    else
+        "in progress"
+    end
+    print(io, typeof(async_result), " (", status, ")")
+end
+
+"""
+    handle_result(async_result::AsyncResult; throw_error=true) -> Result
+
+Executes the query in `async_result` and waits for results.
+
+This implements the loop described in the PostgreSQL documentation for
+[Asynchronous Command Processing](https://www.postgresql.org/docs/10/libpq-async.html).
+
+The `throw_error` option only determines whether to throw errors when handling the new
+[`Result`](@ref)s; the `Task` may error for other reasons related to processing the
+asynchronous loop.
+
+The result returned will be the [`Result`](@ref) of the last query run (the only query if
+using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException`.
+"""
+function handle_result(async_result::AsyncResult; throw_error=true)
+    errors = []
+    result = nothing
+    for result_ptr in _consume(async_result.jl_conn)
+        try
+            result = handle_result(
+                Result(
+                    result_ptr,
+                    async_result.jl_conn;
+                    async_result.result_kwargs[]...
+                );
+                throw_error=throw_error,
+            )
+        catch err
+            push!(errors, err)
+        end
+    end
+
+    if throw_error && !isempty(errors)
+        throw(CompositeException(errors))
+    elseif result === nothing
+        error(LOGGER, "Async query did not return result")
+    else
+        return result
+    end
+end
+
+function _consume(jl_conn::Connection)
+    async_result = jl_conn.async_result
+    result_ptrs = Ptr{libpq_c.PGresult}[]
+    watcher = FDWatcher(socket(jl_conn), true, false)  # can wait for reads
+    try
+        while true
+            if async_result.should_cancel
+                debug(LOGGER, "Received cancel signal for connection $(jl_conn.conn)")
+                _cancel(jl_conn)
+            end
+            debug(LOGGER, "Waiting to read from connection $(jl_conn.conn)")
+            wait(watcher)
+            debug(LOGGER, "Consuming input from connection $(jl_conn.conn)")
+            success = libpq_c.PQconsumeInput(jl_conn.conn) == 1
+            !success && error(error_message(jl_conn))
```

Comment:
Should we run that error though the logger?
```suggestion
            !success && error(Logger, error_message(jl_conn))
```